CenterForOpenScience/osf.io
# QuickFilesNode model and API Views, methods, and tests [OSF-8116] [OSF-8117]

## Purpose

From the [Modeling section of the requirements doc](https://docs.google.com/document/d/1pEmGrsLlpklwaNkbyY8uGzmZX4jjLuDftDI2mr1l4PE/edit#heading=h.qqejb3a5h72):
New QuickFiles model that is an AbstractNode subclass. Here are the parts that are model-specific:
- Cannot delete
- Public upon creation; cannot make private
- Cannot be forked
- Cannot be registered
- Cannot have VOLs
- Cannot have addons other than osfstorage
- GUIDs longer for these nodes (10 chars) 
- When an OSFUser is created, create a QuickFiles node (similar to how a Bookmark collection is created for every user)
- QuickFiles nodes have the title "<<user's fullname>>'s Quick Files"
- If a user updates their name, the title of Quickfiles will change as well.

Also includes API updates: `/v2/users/<uid>/files/` and links on the user detail page to view the files in quickfiles and upload a new one.

## Changes

- New `QuickFilesNode` model and method overrides that match the requirements described above.
- Create a new `QuickFilesNode` for a user when they are created
- Tests for `QuickFilesNode` methods and creation
- Migration to add `QuickFilesNode` model
- Script to add a `QuickFilesNode` for every user that doesn't yet have one
- If a user changes their name, the title of the QuickFiles project changes as well

API side:
- new `/v2/users/<uid>/files/` url and views (also supports `/v2/users/me/files/`)
- files relationship to the user serializer including an upload link for the quickfiles node on waterbutler

## Side effects
- A *long* Migration to add `QuickFilesNode` model and to add a `QuickFilesNode` for all existing users will need to be run
- this PR doesn't include associated frontend work, so all testing for now will need to happen via API.

## Ticket
backend:
https://openscience.atlassian.net/browse/OSF-8116
api:
https://openscience.atlassian.net/browse/OSF-8117

## QA Considerations
- Most of this is backend and API work, so you shouldn't have any user facing UI things to test yet, only API things
- There is a script that needs to be run before existing users will have a quickfiles node, so API things will not work with existing users on staging until that happens
- When testing via the API, posting to the WaterButler `upload` link found at `v2/users/<uid>/` under the files relationship should indeed create new files in your quickfiles "project"
- a GET request on `v2/users/<uid>/files/` should show the quickfiles for that user
- a GET request on `v2/users/me/files/` should show the files for the currently logged in user
